[DataGrid] Add demo page

DanailH opened the pull request:
Fixes #632 

For the Streaming demo that may not be the best solution so I'm opened to suggestions. I took the example from the storybook streaming demos.

https://deploy-preview-1147--material-ui-x.netlify.app/components/data-grid/demo/

dtassone commented on docs/src/pages/components/data-grid/demo/StreamingDemo.js:
```diff
@@ -0,0 +1,186 @@
+import { takeUntil, tap, delay, flatMap } from 'rxjs/operators';
+import { XGrid, useGridApiRef } from '@material-ui/x-grid';
+
+const currencyPairs = [
```
that demo should not be necessary

dtassone commented:
I think we can drop the streaming demo and focus on the main demo with  employee and commodity only.
Once this one is nice we can see how we can integrate real time updates

oliviertassinari commented on docs/src/pages.js:
```diff
@@ -26,6 +26,7 @@ datagrid.children =
           title: 'Overview',
         },
         { pathname: '/components/data-grid/getting-started' },
+        { pathname: '/components/data-grid/demo' },
```
I can't remember what we said about this. Should "Demo" be before "Getting Started" or after?

DanailH replied:
I think we can move it above the "Getting Started" section, makes more sense to check the features before deciding to implement the grid.

oliviertassinari commented on docs/src/pages/components/data-grid/demo/FullFeauredDemo.tsx:
```diff
@@ -0,0 +1,153 @@
+  },
+});
+
+const SettingsPanel = ({ onApply, type, size }) => {
```
While we might not have a convention for the source, for the docs, we are always using this structure:

```suggestion
function SettingsPanel(props) {
  const { onApply, type, size } = props;
```

oliviertassinari commented on docs/src/pages/components/data-grid/demo/FullFeauredDemo.tsx:
```diff
@@ -0,0 +1,153 @@
+      pageSize: settings.pagesize > 0 ? settings.pagesize : undefined,
+    };
+
+    setPagination((p: any) => {
```
no shorthands

oliviertassinari commented on docs/src/pages/components/data-grid/demo/FullFeauredDemo.tsx:
```diff
@@ -0,0 +1,153 @@
+      setType(settings.type.toLowerCase());
+    }
+
+    const newData = await getRealData(
+      settings.size,
+      settings.type === 'Commodity' ? getCommodityColumns() : getEmployeeColumns(),
+    );
+    setGridData(newData);
```
This is unsafe. It could lead to setState called in an unmounted component

DanailH replied:
This should handle the case, I didn't want to wrap it in a cantabile promise.

oliviertassinari commented on docs/src/pages/components/data-grid/demo/FullFeauredDemo.tsx:
```diff
@@ -0,0 +1,153 @@
+
+  return (
+    <div className={classes.root}>
+      <SettingsPanel onApply={onApplyClick} size={size} type={type} />
```
A coding convention

```suggestion
      <SettingsPanel onApply={handleApplyClick} size={size} type={type} />
```

oliviertassinari commented on docs/src/pages/components/data-grid/demo/FullFeauredDemo.tsx:
```diff
@@ -0,0 +1,153 @@
+    width: '100%',
+    '& .MuiFormGroup-options': {
+      alignItems: 'center',
+      justifyContent: 'center',
```
It would look better, I think without centering. Please let me know what you think about it. 
```suggestion
      marginTop: theme.spacing(2),
```

<img width="636" alt="Screenshot 2021-03-02 at 20 32 33" src="https://user-images.githubusercontent.com/3165635/109704055-74cf8500-7b96-11eb-970b-dc097ce36be7.png">

oliviertassinari commented on docs/src/pages/components/data-grid/demo/FullFeauredDemo.tsx:
```diff
@@ -0,0 +1,153 @@
+    '& .MuiFormGroup-options': {
+      alignItems: 'center',
+      justifyContent: 'center',
+      marginBottom: 15,
```
```suggestion
      marginBottom: theme.spacing(2),
```

oliviertassinari commented on docs/src/pages/components/data-grid/demo/FullFeauredDemo.tsx:
```diff
@@ -0,0 +1,153 @@
+      justifyContent: 'center',
+      marginBottom: 15,
+      '& fieldset': {
+        padding: `0 10px`,
```
```suggestion
        padding: theme.spacing(0, 1),
```

oliviertassinari commented on docs/src/pages/components/data-grid/demo/FullFeauredDemo.tsx:
```diff
@@ -0,0 +1,153 @@
+        padding: `0 10px`,
+      },
+      '& button': {
+        marginLeft: 20,
```
```suggestion
        marginLeft: theme.spacing(3),
```

oliviertassinari commented on docs/src/pages/components/data-grid/demo/FullFeauredDemo.tsx:
```diff
@@ -0,0 +1,153 @@
+  root: {
+    display: 'flex',
+    flexDirection: 'column',
+    height: 400,
```
It's not enough, I think that this would start to be interesting:
```suggestion
    height: 600,
```

DanailH replied:
Initially, I set it to 600 but with the 2 demos on the screen, it is a bit too much so I reverted it back. If we are to keep only the full-featured one 600 is definitely better.

oliviertassinari commented on docs/src/pages/components/data-grid/demo/FullFeauredDemo.tsx:
```diff
@@ -0,0 +1,153 @@
+  const [typeState, setType] = React.useState(type);
+  const [selectedPaginationValue, setSelectedPaginationValue] = React.useState(-1);
+
+  const onSizeChange = React.useCallback((e) => {
```
no shorthands
```suggestion
  const onSizeChange = React.useCallback((event) => {
```

oliviertassinari commented on docs/src/pages/components/data-grid/demo/FullFeauredDemo.tsx:
```diff
@@ -0,0 +1,153 @@
+import { XGrid, GridToolbar } from '@material-ui/x-grid';
+import {
+  useDemoData,
+  getCommodityColumns,
+  getEmployeeColumns,
+  getRealData,
```
How about we only use `useDemoData` (or lazy generate the new data, for later :p)?

```suggestion
```

`useDemoData` is optimized to generate data efficiently. Here is the current lag when picking the 100,000 page size option. It's over 5s (in production), it's not an option.

<img width="463" alt="Screenshot 2021-03-02 at 20 39 15" src="https://user-images.githubusercontent.com/3165635/109704850-603fbc80-7b97-11eb-94ef-f4eaed559aa8.png">

Damien has built a static .json, it loads fast enough, loading 1,000 rows. However, it's the same data duplicated x100 to reach 100,000 rows. It's doesn't yield a great sense of randomness in the dispalyed information:

<img width="538" alt="Screenshot 2021-03-02 at 20 42 44" src="https://user-images.githubusercontent.com/3165635/109705437-11deed80-7b98-11eb-89ea-b6bbdb591e05.png">

It would be interesting to see how ag-Grid solve this problem with their large demos

DanailH replied:
So how about we solve the issue the following way -> remove the options bar at the top, keep the data set as "Commodity", make the rows be "more" by default, and set default pagination. That way we are kind of showing the most complete state of the grid. If someone wants to play more they can always open the codesandbox and change the settings (another benefit will be that the demo code will be shorter).

oliviertassinari replied:
I personally think that it's important we show:

- we can handle a **lot** of data, 100,000 or even 1,000,000 would be ideal. With 1,000,000 we will definitely need to lazy load them. It might also hit the max scrollbar height limit.
- we support multiple column types. Having multiple data set can help achieve this goal, but not strong point of view.
- we support the most common use cases, like trading commodities or employes data sets, so two type of data sets makes sens

DanailH replied:
Ok how about this: we keep it as it is now but introduce a loader (it will also showcase the grid loader feature) and I'll add a theme switcher (I think there is one in the demo app)?

oliviertassinari replied:
There is already a dark/light theme switch. I don't think that the demo should have one.

For the loading of data, I would personally rather have 10,000 rows than showing a 5s complet freeze of the JavaScript thread. ag-grid takes a while to load the rows but it's not freezing which is a major difference.

Now, I don't understand why only using `useDemoData` only isn't simpler. For showijg a spinner, we can easily add a hook that defers updates

DanailH replied:
I solved the freeze, it is way more responsive now (only using the `useDemoData` + its API)

DanailH commented:
Thanks, @oliviertassinari and @dtassone  for the review, I'll fix all of those in bulk, remove the streaming demo (I added it because it was mentioned in the ticket), and drop the `demo-app` folder.

oliviertassinari reviewed (commented):
The direction it's moving into seems great 👍. Seems that seems left:

- 100,000 rows performance issue (5s feels really slow). 
- full-screen expansion button

DanailH commented:
> The direction it's moving into seems great 👍. Seems that seems left:
> 
> * 100,000 rows performance issue (5s feels really slow).
> * full-screen expansion button

@oliviertassinari can you check my reply for your comment on the performance issue or I can also remove the option of 100,000 rows for now at least.

oliviertassinari commented:
> About the full-screen button

Ok, no strong preferences. Fixing it will require to update the docs structure of v4 and to forward port it to v5. It's a bit involving, but nothing crazy complex.

oliviertassinari commented on docs/src/pages/components/data-grid/demo/FullFeauredDemo.tsx:
```diff
@@ -0,0 +1,148 @@
+  },
+}));
+
+function SettingsPanel(props) {
+  // eslint-disable-next-line react/prop-types
```
We could potential use TypeScript here, but no real perferences

DanailH replied:
That one actually is not needed. I do need to generate the propTypes tho because the build is failing because of this.

oliviertassinari replied:
> I do need to generate the propTypes 

Well, it's not directly you that generate them, but the script that generate the JS demos from the TS one :)

DanailH replied:
Yes, If the code is ok I'll go ahead and fix/type the `tsx` file.

oliviertassinari replied:
Things I have noticed:

- Ant Design is not spelled correctly
- On mobile, the form field are too close to each other
- It could be interesting to enable the pagination, with pages of 1,000 rows, not sure, only an idea.
- In the future we can improve the useDemoData hook. Have it create the records during the idle time, not blocking the JavaScript thread, as this makes the task asynchronous, add a loading state (loading could be a skeleton, it doesn't have to be circular progress).

It would be great to have @dtassone review on this.

dtassone replied:
the issue with yarn lint is due to the regressions tests, just need to exclude the folders... I put a fix in my autopagesize pr

oliviertassinari replied:
> the issue with yarn lint is due to the regressions tests, just need to exclude the folders... I put a fix in my autopagesize pr

If have left a review on the changes. IMHO they could have been in a standalone PR.

I personally never run eslint locally, sorry, I didn't catch it.

dtassone replied:
I created a PR

dtassone replied:
rebase and merge to fix the eslint issue

oliviertassinari reviewed (commented):
It's a step forward 👍 

I suspect we can improve `useDemoData` to generate the data in chunks, to not freeze the JavaScript thread. Maybe `requestIdleCallback ` would do it. I think that it could also be interesting to update the helper to expose a loading state and bind it to the data grid. But it's likely a problem with a larger scope than #632.

DanailH commented:
@oliviertassinari I added an Antd Design theme switcher (swaps the classes for the grid). I took the styles from here https://material-ui.com/components/data-grid/rendering/#customization-example

oliviertassinari commented:
> @oliviertassinari I added an Antd Design theme switcher (swaps the classes for the grid). I took the styles from here https://material-ui.com/components/data-grid/rendering/#customization-example

Oh ok, I was thinking of this as future potential improvements, once we have an official module developers can important. But why not bridge this in the meantime. As long as it's the same structure as the customization demo, we should be able to maintain them easily (copy & paste the changes)

dtassone commented on docs/src/pages/components/data-grid/demo/FullFeauredDemo.tsx:
```diff
@@ -0,0 +1,270 @@
+  );
+}
+
+export default function FullFeaturedDemo() {
```
Can't this file be split like any applications?
It will be difficult to scale if we have to put everything in one file :D

oliviertassinari replied:
It's possible. If we do so, we should disable the export to codesandbox since imports aren't followed.

DanailH replied:
I had the same thought about this, but then I saw other big demos in the core and I think it is valuable for people to be able to open it directly in **codesandbox**. No preference here - we can keep it all together for now and split them if it becomes too big. Currently, the file is at about 270 lines which is a lot but still acceptable considering that 100 lines are styles.

oliviertassinari replied:
Regarding the option to add multifile support to the codesandbox export, I doubt the ROI is positive.

dtassone replied:
It's  a big demo, not sure this one needs to be open in codesandbox

DanailH replied:
But then how would they copy/paste it or play with the code easily?

oliviertassinari replied:
To add on Damien points, we often display access to the source in the main documentation for playgrounds. It's usually not really useful.

Maybe we can start with a single file, after a couple of days, look at the Google Analytics events to see if developers care about accessing the source?

dtassone commented on docs/src/pages/components/data-grid/demo/demo.md:
```diff
@@ -0,0 +1,14 @@
+
+<p class="description">Use the demos below to explore the available features</p>
+
+## Full featured
```
useless block I think

DanailH replied:
I added it because initially there were 2 demos but for now, maybe we can remove it and add those headers once we have 1+ demos on the page (also the right nav is gone).

dtassone commented:
Great progress 👍 
I think the text below is useless. Just the title is enough
```
Full featured
This demo includes all of the currently available features in XGrid. Use the settings to alter the data and play with the grid.
```

- The demo should be bigger
- Ant theme has no border in dark mode

DanailH commented:
Ok, I fixed the layout on mobile (there is an issue but it is related to the toolbar buttons), fixed the typo in the Ant Design and added the correct border around the grid for the Ant Desing theme.

I also updated the doc page, removed the text.

oliviertassinari reviewed (approved):
We also need to update the navigation in the main repo. In the future, we could solve this side nav duplication issue.

oliviertassinari commented:
You were too eager when removing dependencies. We depend on one for building the visual regression bundle.

oliviertassinari commented on docs/src/pages/components/data-grid/demo/demo.md:
```diff
@@ -0,0 +1,10 @@
+
+<p class="description">Use the demo below to explore the available features</p>
+
+{{"demo": "pages/components/data-grid/demo/FullFeauredDemo.js", "bg": "inline"}}
```
We don't need to show the inline preview, it seems to be noise, we can force it to be collapsed.

DanailH replied:
I don't have a preference. For not it may not be a problem but once there are more demos it will be I think.

oliviertassinari replied:
OK, let's hide the inline preview then, it seems to be a distraction for developers.

DanailH replied:
done

oliviertassinari commented on docs/src/pages/components/data-grid/demo/FullFeauredDemo.js:
```diff
@@ -0,0 +1,282 @@
+          <MenuItem value="Commodity">Commodity</MenuItem>
+        </Select>
+      </FormControl>
+      <FormControl size="small" component="fieldset">
+        <FormLabel component="legend">Rows</FormLabel>
+        <Select value={sizeState} onChange={handleSizeChange}>
+          <MenuItem value={100}>100</MenuItem>
+          <MenuItem value={1000}>{Number(1000).toLocaleString()}</MenuItem>
+          <MenuItem value={10000}>{Number(10000).toLocaleString()}</MenuItem>
+          <MenuItem value={100000}>{Number(100000).toLocaleString()}</MenuItem>
+        </Select>
+      </FormControl>
```
This select component snippet is wrong. I don't know where it's coming from but no. We don't document this.

DanailH replied:
I took the bar from the `demo-app`. I can rework the snippet to be more polished.

oliviertassinari replied:
The errors I have seen:

1. FormLabel instead of InputLabel
2. Not correct component structure, should be a label
3. Flickering bug in Chrome which can from 2.
4. SSR unstable usage of toLocaleString. But it doesn't seem to really be an issue because we don't format the default option. It's inconsistent but an interesting tradeoff, feel free to ignore this point.

DanailH replied:
Updated

DanailH commented:
> You were too eager when removing dependencies. We depend on one for generating the visual regression bundle.

If you are talking about the changes in the `yarn.lock` file what I did is run `yarn install` after removing the demo-app folder. I can bring it back if needed.

oliviertassinari commented:
@DanailH not really, I simply wanted to say that the visual regression tests are broken because there is a missing dependency.

oliviertassinari commented on docs/src/pages/components/data-grid/demo/FullFeauredDemo.tsx:
```diff
@@ -0,0 +1,304 @@
+
+  return (
+    <FormGroup className="MuiFormGroup-options" row>
+      <FormControl component="fieldset">
```
```suggestion
      <FormControl>
```

x the other, this is not a fieldset, it would need a legend to be correct

DanailH replied:
removed

oliviertassinari commented on docs/src/pages/components/data-grid/demo/FullFeauredDemo.tsx:
```diff
@@ -0,0 +1,304 @@
+        theme={getActiveTheme()}
+      />
+      <XGrid
+        className={isAntDesign ? AntDesignClasses.root : undefined}
```
Interesting, if we want to push the approach further. We would need to use the `ThemeProvider`. For instance, would allow to customize the component deeper, like the typography or the button styles

<img width="442" alt="Screenshot 2021-03-05 at 15 19 59" src="https://user-images.githubusercontent.com/3165635/110127773-5059ef80-7dc6-11eb-9c79-58d4aedd714f.png">

@dtassone any preference for when we take on this problem? I think that it could be later, but no strong preferences.

DanailH replied:
I agree, ideally should be with the theme provider but nailing the styles down would take more time.

dtassone replied:
@oliviertassinari let's create a separate issue on prioritize in the next planning session :)

dtassone commented on docs/src/pages/components/data-grid/demo/FullFeauredDemo.js:
```diff
@@ -0,0 +1,281 @@
+import * as React from 'react';
```
filename type

DanailH replied:
Fixed

dtassone commented:
I think it looks great for a first version but overall, I would have expected a bigger demo that looks more like a real-world app...
IMHO it looks pretty much like any other demo that we have 🤔 

Maybe we can reiterate at a later stage

DanailH commented:
> I think it looks great for a first version but overall, I would have expected a bigger demo that looks more like a real-world app...
> IMHO it looks pretty much like any other demo that we have 🤔
> 
> Maybe we can reiterate at a later stage

I think we should have defined it more explicitly because this is a full-feature demo in the sense that all of the grid's functionality is enabled. In the future, we can do what ag-grid are doing with their custom-built navigations that make changing the grid state easier, like a separate filter panel, separate column selector and etc.

dtassone commented:
> > I think it looks great for a first version but overall, I would have expected a bigger demo that looks more like a real-world app...
> > IMHO it looks pretty much like any other demo that we have 🤔
> > Maybe we can reiterate at a later stage
> 
> I think we should have defined it more explicitly because this is a full-feature demo in the sense that all of the grid's functionality is enabled. In the future, we can do what ag-grid are doing with their custom-built navigations that make changing the grid state easier, like a separate filter panel, separate column selector and etc.

I agree that all of this is ok and that it is a great start. Good work!

All I'm saying is that we should think of a way of presenting our grid in this demo in a nicer way...
Real time updates will make things better, but it would be great to improve the layout
